### PR TITLE
update eurac use case

### PR DESCRIPTION
**Pull request by iacopoff**
# Summary

This PR updates the Eurac use case pipeline to run with the refactored version of Hython.

Among other things, it adds the possibility to train a model with a probabilistic head layer, for uncertainty estimation.

The convlstm pipeline has been removed, and it will be reintegrated in the future.

---

**Related issue :**

### CODE REVIEW

File: use-cases/eurac/config.yaml [hunk 1]
```diff
@@ -10,65 +10,195 @@
...
+
+seq_length: 120
+
+learning_rate: 0.001
+
+batch: 256
+
+epochs: 30
+
+gradient_clip: null
```
Comment (jarlsondre) [inline use-cases/eurac/config.yaml:68]: This is not super important, but I feel like the space between each line here makes this a little harder to read. I like the idea of organizing the file into different categories a lot, though, such as "Training".

File: use-cases/eurac/data.py [hunk 2]
```diff
@@ -1,279 +1,56 @@
...
+        scaling_variant: str,
+        experiment_name: str,
+        experiment_run: str,
+        data_source: dict,
+        work_dir: str,
+        dynamic_inputs: list[str] = None,
+        static_inputs: list[str] = None,
+        target_variables: list[str] = None,
+        scaling_static_range: dict = None,
+        mask_variables: list[str] = None,
```
Comment (jarlsondre) [inline use-cases/eurac/data.py:28]: I see two things that I think might be important to consider here: 

1) I would rather use `List` from the `typing` module than `list` here for the sake of conforming to normal Python convention. The same goes for `dict`; I would rather use `Dict`. 
2) These variables all have a value of `None`, but this is never indicated in the type hint. When you say that e.g. `dynamic_inputs` is of the type `list[str]`, the normal Python convention is that it is not possible for `dynamic_inputs` to have the value of `None`. However, the default argument is `None`, which would be a contradiction. To allow for it to be `None` as well, you would use the syntax `list[str] | None`.

File: use-cases/eurac/data.py [hunk 3]
```diff
@@ -1,279 +1,56 @@
...
+        valid_temporal_range: list[str] = ["", ""],
+        train_downsampler: dict = None,
+        valid_downsampler: dict = None,
+        # == calibration ==
+        data_dynamic_inputs: str = None,
+        data_static_inputs: str = None,
+        data_target_variables: str = None,
+        data_target_mask: str = None,
+        min_sample_target: int = None,
+        seq_length: int = None,
```
Comment (jarlsondre) [inline use-cases/eurac/data.py:39]: Same as above with the `None` typing.

File: use-cases/eurac/data.py [hunk 4]
```diff
@@ -1,279 +1,56 @@
...
+        work_dir: str,
+        dynamic_inputs: list[str] = None,
+        static_inputs: list[str] = None,
+        target_variables: list[str] = None,
+        scaling_static_range: dict = None,
+        mask_variables: list[str] = None,
         train_temporal_range: list[str] = ["", ""],
-        test_temporal_range: list[str] = ["", ""],
-        name: str | None = None,
+        valid_temporal_range: list[str] = ["", ""],
```
Comment (jarlsondre) [inline use-cases/eurac/data.py:30]: This might be fine in this specific application, but usually setting a list as a default value in the arguments is considered an anti-pattern in Python. This is because it is specifically bug prone, as the expression is only evaluated once. You can see an example of what I mean in this image: 
image
In this case, you can see that even though I call the function twice, the same dictionary remains. In your code the same will be the case. 

To avoid this, typically you would set the type to "List[str] | None = None" and then do a check in the code where you say
```python
if valid_temporal_range is None:
  valid_temporal_range = ["", ""]
```
This works because this is evaluated separately every time you call the function.

File: use-cases/eurac/trainer.py [hunk 5]
```diff
@@ -104,64 +99,118 @@ def __init__(
...
+            )
 
-        target_weights = {
-            t: 1 / len(self.config.target_names) for t in self.config.target_names
-        }
-        self.loss_fn = RMSELoss(target_weight=target_weights)
-        self.metric_fn = MSEMetric()
+            surrogate.load_state_dict(
+                torch.load(
+                    f"{self.config.work_dir}/{self.config.model_head_dir}/{self.config.model_head_file}"
```
Comment (jarlsondre) [inline use-cases/eurac/trainer.py:151]: In general, it is an advantage to separate these paths out into variables of their own. This is because it makes debugging easier and it allows it to be self-documented since you can give the variable an explanatory name. Additionally, I would use `pathlib`'s `Path` here, as that is considered best practice in modern Python.

File: use-cases/eurac/trainer.py [hunk 6]
```diff
@@ -104,64 +99,118 @@ def __init__(
...
-    @profile_torch_trainer
+    
     @measure_gpu_utilization
     def train(self):
-        """Override version of hython to support distributed strategy."""
+        """Override train_val version of hython to support distributed strategy."""
+
         # Tracking epoch times for scaling test
         if self.strategy.is_main_worker:
             num_nodes = int(os.environ.get("SLURM_NNODES", "unk"))
```
Comment (jarlsondre) [inline use-cases/eurac/trainer.py:212]: In this case you will run into trouble if "SLURM_NNODES" does not exist, as calling `int()` on the string "unk" will fail and thus crash the code.

File: use-cases/eurac/trainer.py [hunk 7]
```diff
@@ -240,12 +270,22 @@ def train(self):
...
-            avg_metrics = pd.DataFrame(metric_history).mean().to_dict()
+            metric_history_ = {}
+            for period in metric_history:
+                for target in metric_history[period]:
+                    for key, value in target.items():
+                        l = []
+                        k = key.lower().split("metric")[0]
+                        n = period + "_" + k
+                        l.append(value)
+                        metric_history_[n] = l
```
Comment (jarlsondre) [inline use-cases/eurac/trainer.py:286]: In general, you want to avoid single-letter variable names such as `l`, `k` and `n`, because it makes it very hard to understand what is going on. Generally, the cost of typing out an explanatory variable name is far far lower than the cost of trying to rewrite this code after forgetting what you were thinking when making this.

**Review by jarlsondre — CHANGES_REQUESTED** [review]
From what I can tell, this looks mostly good. There are a couple of stylistic changes that could be considered, however. Personally, I would install `ruff` or `black` and utilize those to fix some of the stylistic stuff. Also, there is a flag you can add to `ruff` that makes it sort the imports as well, making the code extra clean. If you run `ruff` while inside the `itwinai` project, it will use our own configuration—which you can find in the `pyproject.toml`—that includes sorting the imports. Let me know if anything is unclear 😄

File: use-cases/eurac/config.yaml (hunk 1 above)
Comment (iacopoff) [reply use-cases/eurac/config.yaml]: @jarlsondre one way to organize better the config would be grouping sections into dict as:

```yaml
training:
    strategy: ddp
    hython_trainer: rnntrainer
    loss_fn:
      _target_: hython.losses.RMSELoss
    metric_fn:
      _target_: hython.metrics.MSEMetric
....
model:
   model_file_name: model.pt
   model: cudalstm
   hidden_size: 64
   dropout: 0.1
   model_head_layer: regression # distr_normal
   model_head_activation: linear
   model_head_kwargs: null
...
```
Another advantage is that in this way I can avoid prefixing keys with the section name to avoid potential conflicts with other keys. 

I will think about it, but would be interested in your opinion on that.

**Review by jarlsondre — APPROVED** [review]
Tested training with the new setup and it seems to work 👍
